torch/_dynamo/variables/misc.py

```diff
@@ -234,11 +258,75 @@ def visit(node):
 
         VariableTracker.apply(visit, (args, kwargs))
 
+        ctx = AutogradFunctionContextVariable.create(tx)
+        args = [ctx, *args]
+
         if requires_grad and torch.is_grad_enabled():
-            # TODO(jansel): handle this in training mode
-            unimplemented("autograd.Function with requires_grad")
+            # Note - this is the same check used in autograd/function.py, except inverted.
+            # If we want to support functorch transforms here, we will need to enable this.
+            if (
+                self.fn_cls.setup_context
+                != torch.autograd.function._SingleLevelFunction.setup_context
+            ):
+                unimplemented(
+                    "NYI - autograd.Function with custom setup_context method"
+                )
+
+            vjp_fn = self.fn_cls.vjp  # type: ignore[attr-defined]
+            if vjp_fn is not torch.autograd.Function.vjp:
+                unimplemented("NYI - User defind vjp")
+
+            from .torch import safe_or_raise_always_restore, TorchHigherOrderOperator
+
+            trampoline_autograd_apply = produce_trampoline_autograd_apply(self.fn_cls)
+            trampoline_autograd_fwd = produce_trampoline_autograd_fwd(self.fn_cls)
+            trampoline_autograd_bwd = produce_trampoline_autograd_bwd(self.fn_cls)
+
+            # NOTE [On Tracing autograd.Function w/ grad]
```

Comment:
Though, to Alban's point, some elaboration on the design might be helpful for the future. e.g. for handling speculation of a function with a ctx object, we have turned ctx in a proxy and whitelisted ctx.saved_tensors from the free variable lifting mechanism.

Reply:
> Left some comments and minor nits, the two points I want to make sure of is:
> 
> * safety: that we're disallowing all the methods/attrs we should be for a safe backwards trace

Yea, this is tricky, because we want to balance restrictiveness with perf because we are being overcautious. The attrs are not necessarily bad, but control flow on them is. I might make this a config driven list.

> * question about the testing plan: I remember the set_materialize_grads thing ended up exposing a bug where Dynamo would straight up yell instead of gracefully fallback to eager. It would be bad if we shipped this PR and then regressed existing user code that do have autograd.Function in them esp because autograd.Function is a very popular API. Do we have a good set of test cases for autograd.Function? (like, in torchbench?)

I don't know! More tests is better though, I am happy to add more tests. Fwiw - benchmark models have autograd.Function in them, and they pass, which is a good representation of reality. 

>   * Related: is it worth adding a dynamo config flag to turn off this feature? If config is True, then always graph break on autograd.Function. Or maybe that is just torch._dynamo.disallow_in_graph, or torch._dynamo.disable?

I can add a kill flag for this, on by default. 

> Otherwise, code LGTM

Thanks, responded to comments.